### DISCUSSION TITLE
Migration from psycopg2 - registering new type

### DISCUSSION THREAD

**Discussion by j-adamczyk** [psycopg/psycopg#329] [Q&A] [answered]
In `psycopg2` I have:
```
def cast_psql_date_to_datetime(value, conn):
    if not value:
        return None
    elif value == "infinity":
        return arrow.Arrow.max
    elif value == "-infinity":
        return arrow.Arrow.min

    return parser.parse(value)


psycopg2.extensions.register_type(
    psycopg2.extensions.new_type(
        (
            psycopg2.extensions.PYDATETIME.values
            + psycopg2.extensions.PYDATETIMETZ.values
        ),
        "PG_TIMESTAMP_TO_PYTHON",
        cast_psql_date_to_datetime,
    )
)
```

How can I port this to `psycopg3`? Should I use something like [handling infinity date example](https://www.psycopg.org/psycopg3/docs/advanced/adapt.html#example-handling-infinity-date)? I saw [this issue](https://github.com/psycopg/psycopg/issues/227), but I am not sure how to do this. I want to do this globally.

**Comment by dvarrazzo** [comment, marked as answer]
Yes, the infinity date example is a good starting point. Because you want to customize loading data from the database, you will need to create a `Loader` subclass. You can use the `psycopg.types.datetime.TimestamptzLoader` as a base class and use `super().load()` to mange the non-infinity case.

After you have your class, if you want to use your loader globally, you can register it with `psycopg.adapters.register_loader("timestamptz", MyLoader)`.

**Reply by j-adamczyk** [reply]
@dvarrazzo so basically:
```
class ArrowDateLoader(psycopg.types.datetime.TimestamptzLoader):
    def load(self, data):
        if data == b"infinity":
            return arrow.Arrow.max
        elif data == b"-infinity":
            return arrow.Arrow.min
        else:
            return super().load(data)


psycopg.adapters.register_loader("timestamptz", ArrowDateLoader)
```

Is this correct? And nothing more is required, this will just convert all PYDATETIME and PYDATETIMETZ instances? Or do I need to also do this separately with inheritance from `psycopg.types.datetime.TimestampLoader`?

**Comment by dvarrazzo** [comment]
I didn't test it, but it looks correct.

You should do the same with timestamp without time zone separately, because it is implemented in a separate `TimestampLoader` class. You can probably use a mixin to share the code, but it is simple enough that repeating it seems better.

**Reply by j-adamczyk** [reply]
Thank you!

**Comment by j-adamczyk** [comment]
Posting full migrated code, in case someone will need it in the future:
```python
from psycopg.types.datetime import TimestampLoader, TimestamptzLoader


class ArrowTimestampMixin:
    def load(self, data):
        if data == b"infinity":
            return arrow.Arrow.max
        elif data == b"-infinity":
            return arrow.Arrow.min
        else:
            # inherit from psycopg Loader class that defined load for timestamps
            return super().load(data)


class ArrowTimestampLoader(ArrowTimestampMixin, TimestampLoader):
    pass


class ArrowTimestamptzLoader(ArrowTimestampMixin, TimestamptzLoader):
    pass


psycopg.adapters.register_loader("timestamp", ArrowTimestampLoader)
psycopg.adapters.register_loader("timestamptz", ArrowTimestamptzLoader)
```

EDIT: implement dvarazzo's suggested changes

**Reply by dvarrazzo** [reply]
You are missing the use of the mixin, and can use ` ```python` to colour the syntax.

**Reply by j-adamczyk** [reply]
Thank you, I edited my comment
